Feature/#5 search domain create

maengjjin opened the pull request:
## 개요
- 사용자 위치기반 카테고리, 음식점 검색 api 구현
- 단위 테스트 구현


## 작업 사항
- 사용자 위치 기반 검색 api 구현
- PostGIS를 활용한 거리 계산 및 반경 필터링 sql 적용
- 커스텀 @Query를 사용한 검색 테스트


## 리뷰 요청 포인트
- 로직이 자연스러운지(놓친 부분이 없는지) 확인 부탁드리겠습니다
- db사용을 하여 테스트코드에 ActiveProfiles를 추가를 했는데 이 방법 말고 다른 방법이 있을까요?
- 정렬 타입(SearchSort)처럼 비즈니스 로직에 직접적인 영향을 주지 않는 Enum은 컨트롤러에서 요청 파라미터로 전달받고 있습니다.
현재는 컨트롤러에서 해당 값을 검증한 뒤, 문자열(String) 기준으로 쿼리 분기 처리를 하는 방향으로 구현했는데, 이 접근 방식에 대해 확인 부탁드립니다.

## 참고 자료:
https://velog.io/@chltpdus48/PostgreSQL%EB%A1%9C-%EC%9C%84%EC%B9%98-%EB%8D%B0%EC%9D%B4%ED%84%B0-%ED%99%9C%EC%9A%A9%ED%95%98%EA%B8%B0-earthdistance-PostGIS-%EA%B3%B5%EA%B0%84-%EC%9D%B8%EB%8D%B1%EC%8B%B1

Commit:
feat : 카테고리 & 가게 검색 요구사항 구현

Commit:
test : 카테고리 & 가게 검색 요구사항 테스트

Commit:
style: SearchCommand.Create 코드 포매팅 정리

Commit:
test: 카테고리·거리·정렬·페이징 검증 테스트

Commit:
feat: 검색 응답 DTO 추가 및 검색 API 흐름 정리

Commit:
Merge branch 'dev' into feature/#5-search-domain-create

Commit:
Merge branch 'dev' into feature/#5-search-domain-create

Commit:
fix: 테스트코드 메서드 변경

Commit:
feat: SearchController 테스트 및 로컬yml 테스트 설정

Commit:
style: 테스트코드 패키지 구조 이동

Commit:
fix: 테이블명 수정

dain391 commented on src/main/java/com/babgo/application/search/SearchFasade.java:
```diff
@@ -0,0 +1,36 @@
+
+@Service
+@RequiredArgsConstructor
+public class SearchFasade {
```
표기 오타 있습니다!

maengjjin replied:
수정 완료 했습니다 감사합니다 :)

dain391 commented on src/main/java/com/babgo/controller/search/SearchController.java:
```diff
@@ -0,0 +1,39 @@
+        @RequestParam @NotNull double longitude,
+        @RequestParam @NotNull SearchType searchType,
+        @RequestParam @NotNull String keyword,
+        @RequestParam @NotNull SearchSort sort,
```
검증은 request에서 하는게 컨벤션에 맞을 것 같습니다.

maengjjin replied:
수정 완료 했습니다 감사합니다 :)

dain391 commented on src/main/java/com/babgo/controller/search/SearchRequest.java:
```diff
@@ -0,0 +1,62 @@
+
+        public SearchInfo.Create toSearchInfo(){
+
+            return SearchInfo.Create.of(latitude, longitude, searchType.name(), keyword, sort.name(), page, size);
```
문자열 의존도가 높아보여서 enum 자체를 받는 구조로 변경하는건 어떨까요?

maengjjin replied:
Controller에서 요청 값을 검증할 필요가 있어 해당 계층에 Enum을 정의했습니다.
정렬이나 타입과 같은 값들은 도메인 규칙에 해당하지 않기 때문에, 이를 Enum으로 정의할 경우 오히려 확장성이 떨어질 수 있다고 판단했습니다.
따라서 비즈니스 로직에서는 문자열 형태로 처리하도록 구현했습니다.

dain391 replied:
답변 감사합니다. 말씀하신 의도 이해했습니다!

dain391 commented on src/main/java/com/babgo/application/search/SearchFasade.java:
```diff
@@ -0,0 +1,36 @@
+        List<Search> searchList = searchService.getSearch(searchInfo.toCommand());
+
+        List<CreateResult> results = searchList.stream()
+            .map(search -> CreateResult.of(
```
매핑 로직은 searchInfo나 Search 엔티티에 위치시키는 편이 나을 것 같습니다.

maengjjin replied:
수정 완료 했습니다 감사합니다 :)

dain391 commented on src/main/java/com/babgo/domain/search/SearchCommand.java:
```diff
@@ -0,0 +1,47 @@
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class SearchCommand {
```
SearchInfo와 DTO 중복 있는 것 같습니다.

maengjjin replied:
어플리케이션 계층과 도메인 계층 간 DTO 전달을 위해 객체를 생성했습니다.
파사드에서 Info를 Search로 매핑하려 했지만, 해당 DTO는 Search의 일부 필드만 포함하고 있어 직접 매핑하기에는 적합하지 않았습니다.
따라서 이 경우에는 도메인과 1:1 매핑이 어려워 Command 객체를 별도로 생성하는 것이 더 타당하다고 판단했습니다.
또한 파라미터를 개별로 전달하는 방식은 향후 운영 및 확장 시 관리가 복잡해질 수 있으므로, 객체 단위로 처리하는 방식을 선택했습니다.

dain391 replied:
설명 감사합니다. Command를 별도 객체로 둔 이유 이해했습니다!
record를 활용하는 이유가 확실히 느껴지네요...

dain391 commented on src/main/java/com/babgo/controller/search/SearchController.java:
```diff
@@ -0,0 +1,39 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/search")
```
`/v1/search`가 컨벤션에 맞습니다!

maengjjin replied:
수정 완료 했습니다 감사합니다 :)

dain391 commented on src/main/java/com/babgo/domain/search/Search.java:
```diff
@@ -0,0 +1,83 @@
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@Entity
+@Getter
+@Table(name = "p_store_searches")
```
JPA 쿼리와 테이블명이 다릅니다.

maengjjin replied:
수정 완료 했습니다 감사합니다 :)

dain391 commented on src/main/java/com/babgo/domain/search/SearchService.java:
```diff
@@ -0,0 +1,35 @@
+        
+        return switch (searchType) {
+            case "STORE" -> searchByStore(searchCommand);
+            case "KATEGORIE" -> searchByCategory(searchCommand);
```
Enum의 name을 그대로 쓰는데 하드코딩 리스크 없을까요??
Enum 자체를 넘기는게 안전하게 잡힐 것 같아서요!

dain391 commented:
import문 최적화 부탁드립니다!

Commit:
fix: 테이블명, 클래스명, URL, 접근제한자 수정

Commit:
refactor: Facade의 매핑 로직을 Info 객체로 위임

Commit:
refactor: RequestParam 검증을 Request DTO 생성자 검증으로 변경

Commit:
refactor: 검색 타입 분기 로직을 파사드로 이동 및 타입변경

- 검색 타입 분기 로직이 비즈니스 로직이 아니므로 파사드 계층으로 이동
- 검색 타입을 String에서 Enum으로 변경하여 타입 안정성 유지

Commit:
refactor: 가게 생성순 정렬 쿼리 추가 및 테스트 코드 작성

dain391 reviewed (approved):
수고하셨습니다!
